### ISSUE TITLE
Constraint's equality attribute sometimes fails to properly identify equality constraints

### ISSUE THREAD

**Issue by bknueven** [Pyomo/pyomo#2093]
## Summary
Occasionally constraint's new equality attribute will fail in ways that are difficult to reproduce.

### Steps to reproduce the issue
This happened while I was running the sequential decomposition tool, but I can't seem to create a small example which reproduces the issue on my system. Instead I've included some hopefully useful output from the debugger:
```python
ipdb> type(con)
<class 'pyomo.core.base.constraint._GeneralConstraintData'>
ipdb> type(con._expr)
<class 'pyomo.core.expr.logical_expr.RangedExpression'>
ipdb> con.pprint()
{Member of flow_mass_phase_comp_equality} : Size=2, Index=fs.feed._flow_mass_phase_comp_outlet_ref_index, Active=True
    Key                 : Lower : Body                                                                                                                                       : Upper : Active
    (0.0, 'Liq', 'H2O') :   0.0 : fs.feed.properties[0.0].flow_mass_phase_comp[Liq,H2O] - fs.PrimaryPumps[1].control_volume.properties_in[0.0].flow_mass_phase_comp[Liq,H2O] :   0.0 :   True
ipdb> con.lb == con.ub
True
ipdb> con.lb is con.ub
False
ipdb> type(con.lb)
<class 'float'>
ipdb> type(con.ub)
<class 'float'>
ipdb> id(con.lb)
140324945312432
ipdb> id(con.ub)
140324945312400
ipdb> 
```

### Information on your system

Pyomo version: 6.1.1
Python version: 3.7.9
Operating system: Ubuntu 20.04.2
How Pyomo was installed (PyPI, conda, source): source
Solver (if applicable): N/A

### Additional information
I believe the issue is with the `is` check on bounds in the `equality` attribute on line 416:
https://github.com/Pyomo/pyomo/blob/3cf310ac1d52b7e362dc2eaf6b8a59ac4c896fe4/pyomo/core/base/constraint.py#L408-L418

For `RangedExpressions` with floating point bounds, CPython does not guarantee that the same valued `float` objects will occupy the same location in memory (as shown by the debugging output above). However, it is usually the case that identical floating point values *do* occupy the same location in memory, making this difficult to reproduce. There was some discussion around resolving this issue in Python itself: https://discuss.python.org/t/demoting-the-is-operator-to-avoid-an-identity-crisis/86, but the CPython team went a different route: https://bugs.python.org/issue34850.

**Comment by jsiirola** [comment]
I was thinking about this when implementing it.  The question in my mind is if `equality` should be "structurally" or "effectively".  That is, consider:
```python
model.x = Var()
model.l = Param(initialize=0, mutable=True)
model.u = Param(initialize=0, mutable=True)
model.c = Constraint(expr=inequality(model.l, model.x, model.u)
```
Should `model.c.equality` be True or False?  If the answer is `True`, would it be weird if the user changed model.u to be 1 and `model.c.equality` suddenly became `False`?

One alternative could be to only use `is` when `_lower` / `_upper` are non-constant Pyomo objects, and fall back on equality for native times.  I think I like this compromise.

**Comment by bknueven** [comment]
I agree that `is` should be used for non-constant Pyomo objects for exactly your case. So your compromise makes good sense to me.

Note that the original constraint above was created by the `network.expand_arcs` transformation as `var1 == var2`. So in this instance the expression system created this issue by unluckily grabbing two separate instances of `0.`. I might not have noticed except the sequential decomposition tool checks the equality attribute on the expanded arc constraints.

**Comment by jsiirola** [comment]
@bknueven, while #2097 resolves this specific behavior, I am not at all clear how you are getting a `RangedExpression` in your constraint to begin with.  Can you share your model / driver (off-issue is fine)
